grafana/opensearch-datasource
# Refactor ad hoc variable processing

**What this PR does / why we need it**:
Previously, we were interpolating variables in the `query()` method manually. This adds several methods to the datasource that are called by code Grafana whenever variable and ad hoc interpolation is necessary: 

- `applyTemplateVariables` - this is a method that is called by Grafana during the super.query() call, which means we don't have to do this manually in our datasource. The method implements template variable interpolation as well as ad hoc queries following the approach in Elastic search from [this PR](https://github.com/grafana/grafana/pull/65274)
- `interpolateVariablesInQueries` called from multiple places in core grafana, didn't look too much into it but most datasources implement it. Elastic [here](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/elasticsearch/datasource.ts#L431)
- copied processing for ad hoc filters for lucene queries added in Elastic [in this PR](https://github.com/grafana/grafana/pull/71954) 
- added processing for PPL ad hoc filters. This is mostly copied from what we had, with the difference of adding removeFilterFromPPLQuery, PPLQueryHasFilter, toggleQueryFilterForPPL etc. that are now used for both adding ad hoc filters in dashboards and in Explore
- also fixes https://github.com/grafana/opensearch-datasource/issues/389 by [interpolating the stringified query](https://github.com/grafana/opensearch-datasource/blob/1dc25998bdf13e82ad51b2ff05f3ee983e827dec/src/datasource.ts#L412)

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/opensearch-datasource/issues/379, https://github.com/grafana/opensearch-datasource/issues/389

**Special notes for your reviewer**:
- this was mostly tested with backend flow, but the frontend flow seems to be improved as well, mainly when adding ad hoc queries in Explore

**How to test**
1. In Dashboard run Lucene "Raw data" query with table panel and click on magnifying glasses in the table cells. 
- `+` adds a filter
- `+` again removes a filter
- `-` filters out a field.
2. Make sure that executed query (using query inspector) has filters as `query.query`. The query won't have filters in the UI, just the query inspector
3. Switch to Explore and do the filtering in the same way. In Explore, the query should be shown in the UI Query field
4. Repeat 1-3 for PPL queries
